### DISCUSSION TITLE
Dissolution Chamber support for automateable recipes (lockable slots? side config?)

### DISCUSSION THREAD

**Discussion by nerdywhiteguy** [InnovativeOnlineIndustries/Industrial-Foregoing#928] [Completed/Done]
# There is currently no way to automate item crafting in a Dissolution Chamber.

As a fan of "automate all the things!" style modpacks, and having put together a few such modpacks for myself and friends, it saddens me that there is no way to automate a Dissolution Chamber. If you use a hopper or any other item transport, items are placed into the slots in left-to-right, top-to-bottom order. If you can somehow precisely control the order of items going into the Chamber, it could *theoretically* be automated, but I've never been able to figure out how to do this.

I would very much like to use it's products in my complex crafting setups, and perhaps add custom recipes to the Chamber since it supports 8 input items, with a liquid input *and* output. It can make for some very nice and very flavorful recipes! Unfortunately, any recipe I design for it, I know won't be automatable and players will be forced to craft items by hand every time. As such, I won't want to make my modpack heavily dependent on a Dissolution Chamber recipe, which restricts my design choices.

The exact method to automate the recipe could vary, but here are a few ideas:

## Lockable inventory slots
Like the Plant Sower, let items be placed into the slots and then "locked in" so that something like a hopper could feed items into the Chamber and they'll go into the correct slots. Also, empty slots would need to be locked as empty, since the Chamber won't process if there are extra items in unused slots for that recipe.

## Side config
If different slots could be paired to different sides of the machine, this would allow hoppers (or item transport cables like Mekanism's Logistical Transporters) to insert items directly into the correct slot. It would certainly still keep things from being too easy, while allowing complicated builds. The only problem would be that there are only 6 sides of a cube, and there are 8 item inputs, in addition to the liquid input, the item output, and a possible liquid output. Many recipes take two or more of a given item, so those could be the same side, but it does limit things somewhat.

## Mekanism Transporter Config
Mekanism's Logistical Transporters can be color coded with one of 12 colors, and Logistical Sorters can color code items as they are put into the network. This can be used to create complex and multi-tiered sorting systems. Individual machines can then have a Transporter Config which works specifically with the color-coded Logistical Transporters to only accept items that have a given color-code on certain sides, and can even color code the output. Since these are tied to sides (although they could possibly be tied to slots, but I haven't seen that done), it will have the same limits as the previous suggestion about side configs. However, the advantage here is that, unless a mod like Mekanism is installed, Dissolution Chambers will still feel quite "vanilla" and won't be automateable.

## An upgraded "automatic" variant machine
If it's desirable to keep the Dissolution Chamber manual, perhaps a second (more expensive and/or more energy hungry) machine could be made that fills the automatic role.

## An addon card
Like the speed and range upgrades, perhaps a general purpose "Automation Upgrade" could be created. If this allowed locking inventory slots to only take one kind of item in *all* IF machines, it could be a very nice addition! The price could also be moderately high, since it's not *required* to use the machine, but *is* required by people like me who want to build automatic crafting systems. (Although please don't make it *too* expensive, since I'll need a separate Dissolution Chamber and "Automation Upgrade Addon" for each recipe I want to automate! I'd recommend taking inspiration from Mekanism's mid-tier Control Circuits: Iron, Redstone, and Osmium, a fairly common core) Depending on where you want to go with this, you could also make this addon required for the machines to respond to redstone signals, although you could always use redstone to just cut the power since most power cable mods can be redstone controlled.

Looking forward to seeing where the mod goes, and especially this Quality of Life update to hopefully smooth out some of the rough edges. Thanks for all the hard work!

**Comment by bcrotty** [comment]
I second this request. I would love to automate it. I know #745's fix locks the slots while crafting is going on, but we really need locked slots at all times. When I'm trying to create Pink Slimes, as long as there is enough liquid Pink Slime, crafting continues automatically. When I run out of liquid Pink Slime, whether items are being pushed in or pulled in, all 8 slots fill up, and when I get more liquid Pink Slime, crafting doesn't restart.

**Comment by namibj** [comment]
My workaround is via CC: Tweaked, having a computer to the right of the chamber and a chest on top.
It has the input items placed in slots 5-12 of the chest, currently relying on them being stackable. Just call it with a number to tell it how many it should do. Fluid handling is manual, as of now.
```lua
a = peripheral.wrap("left")
b = "top"
for i=1,arg[1] do
    for j=5,12 do
        print(i .. " " .. " " .. a.pullItems(b, j, 1, j))
    end
    while a.pushItems(b, 15, 1, 15) == 0 do
        sleep(0.05)
    end
end
```
Pastebin: https://pastebin.com/gN5UQPDs

I, too, would prefer something nicer, ideally something where one machine could support multiple recipes.

**Comment by kosma** [comment]
Refined Crafting is able to automate Dissolution Chamber even though the patterns don't look correct. It's not perfect, but it's one way to do it.
